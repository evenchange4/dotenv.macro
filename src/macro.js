// @flow
import dotenv from 'dotenv';
import dotenvExpand from 'dotenv-expand';
import { createMacro } from 'babel-plugin-macros';
// import printAST from 'ast-pretty-print';
// console.log(printAST(referencePath.parentPath))

function valueExpression(
  t: any,
  { name, value }: { name: string, value: string },
): any {
  return t.logicalExpression(
    '||',
    t.logicalExpression(
      '&&',
      t.logicalExpression(
        '&&',
        t.identifier('process'),
        t.memberExpression(t.identifier('process'), t.identifier('env')),
      ),
      t.memberExpression(t.identifier('process.env'), t.identifier(name)),
    ),
    t.valueToNode(value),
  );
}

let cacheDotenv;

function dotenvMacro({
  references,
  babel: { types: t },
}: {
  references: { default: any },
  babel: { types: any },
}): void {
  // Note: Load dotenv file
  if (!cacheDotenv) {
    cacheDotenv = dotenv.config();
    dotenvExpand(cacheDotenv);
  }
  const { default: defaultEnvs = [], ...destructureEnv } = references;

  // Case 1: import env from 'dotenv.macro'
  defaultEnvs.forEach(referencePath => {
    if (referencePath.parentPath.type === 'MemberExpression') {
      const { name } = referencePath.parentPath.node.property;
      const value = process.env[name];
      referencePath.parentPath.replaceWith(valueExpression(t, { name, value }));
    } else {
      const parsedAst = Object.keys(cacheDotenv.parsed).map(name => {
        const value = (process.env: any)[name];
        return t.objectProperty(
          t.stringLiteral(name),
          valueExpression(t, { name, value }),
        );
      });
      referencePath.replaceWith(t.objectExpression(parsedAst));
    }
  });

  // Case 2: import { PORT } from 'dotenv.macro'
  Object.values(destructureEnv).forEach((referencePaths: any) => {
    referencePaths.forEach(referencePath => {
      const { name } = referencePath.node;
      const value = (process.env: any)[referencePath.node.name];
      referencePath.replaceWith(valueExpression(t, { name, value }));
    });
  });
}

export default createMacro(dotenvMacro);
