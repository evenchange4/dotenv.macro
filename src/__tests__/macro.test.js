import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

expect.addSnapshotSerializer({
  print(val) {
    return val.replace(/..\/macro/, 'dotenv.macro');
  },
  test(val) {
    return typeof val === 'string';
  },
});

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
  },
  tests: {
    '[import] all': {
      error: false,
      code: `
        import env from '../macro';
        console.log(env);
      `,
    },
    '[import] inline the value': {
      error: false,
      code: `
        import env from '../macro';
        console.log(env.PORT);
      `,
    },
    '[import] destructure with single value': {
      error: false,
      code: `
        import { PORT } from '../macro';
        console.log(PORT);
      `,
    },
    '[import] destructure with multiple values': {
      error: false,
      code: `
        import { PORT, HOST } from '../macro';
        console.log(PORT);
        console.log(HOST);
      `,
    },
    '[require] all': {
      error: false,
      code: `
        const env = require('../macro');
        console.log(env);
      `,
    },
    '[require] inline the value': {
      error: false,
      code: `
        const env = require('../macro');
        console.log(env.PORT);
      `,
    },
    '[require] destructure with single value': {
      error: false,
      code: `
        const { PORT } = require('../macro');
        console.log(PORT);
      `,
    },
    '[require] destructure with multiple values': {
      error: false,
      code: `
        const { PORT, HOST } = require('../macro');
        console.log(PORT);
        console.log(HOST);
      `,
    },
    '[require] NODE_ENV from CLI': {
      error: false,
      code: `
        const { NODE_ENV } = require('../macro');
        console.log(NODE_ENV);
      `,
    },
  },
});
