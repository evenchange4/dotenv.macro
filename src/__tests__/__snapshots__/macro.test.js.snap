// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros [import] all: [import] all 1`] = `

import env from 'dotenv.macro';
console.log(env);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log({
  "PORT": process && process.env && process.env.PORT || "3001",
  "HOST": process && process.env && process.env.HOST || "localhost"
});

`;

exports[`macros [import] destructure with multiple values: [import] destructure with multiple values 1`] = `

import { PORT, HOST } from 'dotenv.macro';
console.log(PORT);
console.log(HOST);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.PORT || "3001");
console.log(process && process.env && process.env.HOST || "localhost");

`;

exports[`macros [import] destructure with single value: [import] destructure with single value 1`] = `

import { PORT } from 'dotenv.macro';
console.log(PORT);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.PORT || "3001");

`;

exports[`macros [import] inline the value: [import] inline the value 1`] = `

import env from 'dotenv.macro';
console.log(env.PORT);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.PORT || "3001");

`;

exports[`macros [require] NODE_ENV from CLI: [require] NODE_ENV from CLI 1`] = `

const { NODE_ENV } = require('dotenv.macro');
console.log(NODE_ENV);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.NODE_ENV || "test");

`;

exports[`macros [require] all: [require] all 1`] = `

const env = require('dotenv.macro');
console.log(env);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log({
  "PORT": process && process.env && process.env.PORT || "3001",
  "HOST": process && process.env && process.env.HOST || "localhost"
});

`;

exports[`macros [require] destructure with multiple values: [require] destructure with multiple values 1`] = `

const { PORT, HOST } = require('dotenv.macro');
console.log(PORT);
console.log(HOST);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.PORT || "3001");
console.log(process && process.env && process.env.HOST || "localhost");

`;

exports[`macros [require] destructure with single value: [require] destructure with single value 1`] = `

const { PORT } = require('dotenv.macro');
console.log(PORT);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.PORT || "3001");

`;

exports[`macros [require] inline the value: [require] inline the value 1`] = `

const env = require('dotenv.macro');
console.log(env.PORT);

      ↓ ↓ ↓ ↓ ↓ ↓

console.log(process && process.env && process.env.PORT || "3001");

`;
